File: Example/MXPagerView/MXPagerViewExample.m
// MXPagerViewExample.m
//
// Copyright (c) 2015 Maxime Epain
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

#import "MXPagerViewExample.h"
#import <MXPagerView/MXPagerView.h>

@interface MXPagerViewExample () <MXPagerViewDelegate, MXPagerViewDataSource, UITableViewDelegate, UITableViewDataSource>
@property (weak, nonatomic) IBOutlet MXPagerView *pagerView;
@property (nonatomic, strong) UITableView   *tableView;
@property (nonatomic, strong) UIWebView     *webView;
@property (nonatomic, strong) UITextView    *textView;
@end

@implementation MXPagerViewExample

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.pagerView.delegate = self;
    self.pagerView.dataSource = self;
    
    self.pagerView.gutterWidth = 20;
    
    //Register UItableView as page
    [self.pagerView registerClass:[UITextView class] forPageReuseIdentifier:@"TextPage"];
}

#pragma mark Properties

- (UITableView *)tableView {
    if (!_tableView) {
        //Add a table page
        _tableView = [[UITableView alloc] init];
        _tableView.delegate = self;
        _tableView.dataSource = self;
    }
    return _tableView;
}

- (UIWebView *)webView {
    if (!_webView) {
        // Add a web page
        _webView = [[UIWebView alloc] init];
        NSString *strURL = @"http://nshipster.com/";
        NSURL *url = [NSURL URLWithString:strURL];
        NSURLRequest *urlRequest = [NSURLRequest requestWithURL:url];
        [_webView loadRequest:urlRequest];
    }
    return _webView;
}

- (UITextView *)textView {
    if (!_textView) {
        // Add a text page
        _textView = [[UITextView alloc] init];
        NSString *filePath = [[NSBundle mainBundle]pathForResource:@"LongText" ofType:@"txt"];
        _textView.text = [[NSString alloc]initWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:nil];
    }
    return _textView;
}

#pragma mark <MXPagerViewDelegate>

- (void) pagerView:(nonnull MXPagerView *)pagerView didMoveToPageAtIndex:(NSInteger)index {
    NSLog(@"Did move to page with index %li", index);
}

#pragma -mark <MXPagerViewDataSource>

- (NSInteger)numberOfPagesInPagerView:(MXPagerView *)pagerView {
    return 10;
}

- (NSString *)segmentedPager:(MXPagerView *)pagerView titleForSectionAtIndex:(NSInteger)index {
    if (index < 3) {
        return [@[@"Table", @"Web", @"Text"] objectAtIndex:index];
    }
    return [NSString stringWithFormat:@"Page %li", (long) index];
}

- (UIView *)pagerView:(MXPagerView *)pagerView viewForPageAtIndex:(NSInteger)index {
    if (index < 3) {
        return [@[self.tableView, self.webView, self.textView] objectAtIndex:index];
    }
    
    //Dequeue reusable page
    UITextView *page = [self.pagerView dequeueReusablePageWithIdentifier:@"TextPage"];
    NSString *filePath = [[NSBundle mainBundle]pathForResource:@"LongText" ofType:@"txt"];
    page.text = [[NSString alloc]initWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:nil];
    
    return page;
}

#pragma mark <UITableViewDelegate>

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    NSInteger index = (indexPath.row % 2) + 1;
    [self.pagerView showPageAtIndex:index animated:YES];
}

#pragma mark <UITableViewDataSource>

- (NSInteger) tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 50;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *CellIdentifier = @"Cell";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier];
    }
    cell.textLabel.text = (indexPath.row % 2)? @"Text": @"Web";
    
    return cell;
}


@end

File: Example-Swift/MXPagerView-Swift/MXPagerViewExample.swift
// MXPagerViewExample.swift
//
// Copyright (c) 2015 Maxime Epain
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

import UIKit
import MXPagerView

class MXPagerViewExample: UIViewController, MXPagerViewDelegate, MXPagerViewDataSource, UITableViewDelegate, UITableViewDataSource {

    @IBOutlet weak var pagerView: MXPagerView!
    
    var tableView: UITableView!
    var webView: UIWebView!
    var textView: UITextView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        //Init pages
        self.tableView = UITableView()
        self.tableView.delegate = self
        self.tableView.dataSource = self
        self.tableView.registerClass(UITableViewCell.self, forCellReuseIdentifier: "Cell")
        
        self.webView = UIWebView()
        let url = NSURL(string: "http://nshipster.com/");
        let request = NSURLRequest(URL: url!);
        self.webView.loadRequest(request);
        
        self.textView = UITextView()
        let filePath = NSBundle.mainBundle().pathForResource("LongText", ofType: "txt")
        self.textView.text = try! String(contentsOfFile:filePath!, encoding: NSUTF8StringEncoding)
        
        //Init title
        self.navigationItem.title = "Page 0"
        
        //Setup pager
        self.pagerView.gutterWidth = 20
        
        //Register UITextView as page
        self.pagerView.registerClass(UITextView.self, forPageReuseIdentifier: "TextPage")
    }

    @IBAction func previous(sender: AnyObject) {
        self.pagerView.showPageAtIndex((self.pagerView.indexForSelectedPage - 1), animated: true)
    }

    @IBAction func next(sender: AnyObject) {
        self.pagerView.showPageAtIndex((self.pagerView.indexForSelectedPage + 1), animated: true)
    }
    
    // MARK: - Pager view delegate
    
    func pagerView(pagerView: MXPagerView, didMoveToPageAtIndex index: Int) {
        self.navigationItem.title = String(format: "Page %li", index)
    }
    
    // MARK: - Pager view data source
    
    func numberOfPagesInPagerView(pagerView: MXPagerView) -> Int {
        return 10
    }
    
    func pagerView(pagerView: MXPagerView, viewForPageAtIndex index: Int) -> UIView? {
        if index < 3 {
            return [self.tableView, self.webView, self.textView][index]
        }
        
        let page = pagerView.dequeueReusablePageWithIdentifier("TextPage") as! UITextView
        let filePath = NSBundle.mainBundle().pathForResource("LongText", ofType: "txt")
        page.text = try! String(contentsOfFile:filePath!, encoding: NSUTF8StringEncoding)
        
        return page
    }

    // MARK: - Table view data source
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 50
    }
    
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell: UITableViewCell! = tableView.dequeueReusableCellWithIdentifier("Cell", forIndexPath: indexPath)

        cell.textLabel?.text = (indexPath.row % 2 > 0) ? "Text" : "Web"
        
        return cell
    }
    
    // MARK: - Table view delegate
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        let index = (indexPath.row % 2) + 1
        self.pagerView.showPageAtIndex(index, animated:true)
    }
}

